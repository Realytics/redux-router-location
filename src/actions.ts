import { Location, Path, LocationDescriptorObject, createLocation } from "history";
import { ActionCreator } from "redux";

/**
 * Types
 */

export type BareAction = {
  type: typeof GO_BACK | typeof GO_FORWARD;
};

export type IndexedAction = {
  type: typeof GO;
  payload: number;
};

export type LocationAction = {
  type: typeof LOCATION_CHANGED | typeof PUSH | typeof REPLACE;
  payload: Location;
};

/**
 * Action
 */

export const LOCATION_CHANGED: "@@ROUTER/LOCATION_CHANGED" = "@@ROUTER/LOCATION_CHANGED";
export const PUSH: "@@ROUTER/PUSH" = "@@ROUTER/PUSH";
export const REPLACE: "@@ROUTER/REPLACE" = "@@ROUTER/REPLACE";
export const GO: "@@ROUTER/GO" = "@@ROUTER/GO";
export const GO_BACK: "@@ROUTER/GO_BACK" = "@@ROUTER/GO_BACK";
export const GO_FORWARD: "@@ROUTER/GO_FORWARD" = "@@ROUTER/GO_FORWARD";

/**
 * Action creators
 */

export const push = (href: Path | LocationDescriptorObject): LocationAction => ({
  type: PUSH,
  payload: createLocation(href),
});

export const replace = (href: Path | LocationDescriptorObject): LocationAction => ({
  type: REPLACE,
  payload: createLocation(href),
});

export const go = (index: number): IndexedAction => ({
  type: GO,
  payload: index,
});

export const goBack = (): BareAction => ({
  type: GO_BACK,
});

export const goForward = (): BareAction => ({
  type: GO_FORWARD,
});

export const locationDidChange = (location: Location): LocationAction => ({
  type: LOCATION_CHANGED,
  payload: location,
});
