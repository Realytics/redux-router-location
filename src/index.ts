export { createReducer, RouterState } from './reducer';
export { createMiddleware } from './middleware';
export { createEnhancer } from './enhancer';
export {
  LOCATION_CHANGED, PUSH, REPLACE, GO, GO_BACK, GO_FORWARD,
  BareAction, IndexedAction, LocationAction,
  push, replace, go, goBack, goForward,
  locationDidChange
} from './actions';