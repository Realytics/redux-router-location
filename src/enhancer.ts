import { StoreEnhancerStoreCreator, Reducer, StoreEnhancer, Store } from 'redux';
import { History } from 'history';
import { locationDidChange } from './actions';

export function createEnhancer(history: History): StoreEnhancer<any> {
  return (next: StoreEnhancerStoreCreator<any>) => (
    userReducer: Reducer<any>,
    initialState: any
  ) => {
    const store: Store<any> = next(
      userReducer,
      initialState
    );

    history.listen(location => {
      store.dispatch(locationDidChange(location));
    });

    return store;
  };
}
