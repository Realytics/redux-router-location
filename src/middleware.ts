import { History } from 'history';
import { Dispatch, Middleware } from 'redux';
import * as Actions from './actions';

export function createMiddleware(history: History): Middleware {
  return () => {
    return (next: Dispatch<any>) => {
      return (action: any) => {
        switch (action.type) {
          case Actions.PUSH:
            history.push(action.payload);
            // No return, no next() here
            // We stop all history events from progressing further through the dispatch chain...
            break;
          case Actions.REPLACE:
            history.replace(action.payload);
            break;
          case Actions.GO:
            history.go(action.payload);
            break;
          case Actions.GO_BACK:
            history.goBack();
            break;
          case Actions.GO_FORWARD:
            history.goForward();
            break;
          default:
            // ...but we want to leave all events we don't care about undisturbed
            return next(action);
        }
      };
    };
  };
}
