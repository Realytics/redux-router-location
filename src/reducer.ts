import * as Actions from './actions';
import { Location } from 'history';
import { Reducer } from 'redux';

export type RouterState = Location;

/**
 * Only LOCATION_CHANGED is handled here because other action change
 * location which dispatch a LOCATION_CHANGED
 */

export function createReducer(initialLocation: Location): Reducer<RouterState> {
  return (state: RouterState = initialLocation, action: any): RouterState => {
    if (action && action.type === Actions.LOCATION_CHANGED) {
      // No-op the initial route action
      if (state && state.pathname === action.payload.pathname && state.search === action.payload.search) {
        return state;
      }

      return action.payload;
    }
    return state;
  };
}
