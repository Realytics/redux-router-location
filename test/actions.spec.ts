import { push, replace, go, goBack, goForward, locationDidChange } from '../src/actions';
import { Location } from 'history';

function createLocation(path: string = ''): Location {
  return {
    pathname: path,
    hash: '',
    key: '0uicnx',
    search: '',
    state: null,
  };
}

describe('actions', () => {

  describe('push', () => {
    it('create an push action', () => {
      expect(push('/home')).toMatchSnapshot();
      expect(push({ pathname: '/home', search: 'test' })).toMatchSnapshot();
    });
  });

  describe('replace', () => {
    it('create an replace action', () => {
      expect(replace('/home')).toMatchSnapshot();
      expect(replace({ pathname: '/home', search: 'test' })).toMatchSnapshot();
    });
  });

  describe('go', () => {
    it('create an go action', () => {
      expect(go(3)).toMatchSnapshot();
    });
  });

  describe('goBack', () => {
    it('create an goBack action', () => {
      expect(goBack()).toMatchSnapshot();
    });
  });

  describe('goForward', () => {
    it('create an goForward action', () => {
      expect(goForward()).toMatchSnapshot();
    });
  });

  describe('locationDidChange', () => {
    it('create an locationDidChange action', () => {
      expect(locationDidChange(createLocation('/home'))).toMatchSnapshot();
    });
  });

});
