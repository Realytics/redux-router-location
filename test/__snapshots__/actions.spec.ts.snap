// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions go create an go action 1`] = `
Object {
  "payload": 3,
  "type": "@@ROUTER/GO",
}
`;

exports[`actions goBack create an goBack action 1`] = `
Object {
  "type": "@@ROUTER/GO_BACK",
}
`;

exports[`actions goForward create an goForward action 1`] = `
Object {
  "type": "@@ROUTER/GO_FORWARD",
}
`;

exports[`actions locationDidChange create an locationDidChange action 1`] = `
Object {
  "payload": Object {
    "hash": "",
    "key": "0uicnx",
    "pathname": "/home",
    "search": "",
    "state": null,
  },
  "type": "@@ROUTER/LOCATION_CHANGED",
}
`;

exports[`actions push create an push action 1`] = `
Object {
  "payload": Object {
    "hash": "",
    "key": undefined,
    "pathname": "/home",
    "search": "",
    "state": undefined,
  },
  "type": "@@ROUTER/PUSH",
}
`;

exports[`actions push create an push action 2`] = `
Object {
  "payload": Object {
    "hash": "",
    "key": undefined,
    "pathname": "/home",
    "search": "?test",
  },
  "type": "@@ROUTER/PUSH",
}
`;

exports[`actions replace create an replace action 1`] = `
Object {
  "payload": Object {
    "hash": "",
    "key": undefined,
    "pathname": "/home",
    "search": "",
    "state": undefined,
  },
  "type": "@@ROUTER/REPLACE",
}
`;

exports[`actions replace create an replace action 2`] = `
Object {
  "payload": Object {
    "hash": "",
    "key": undefined,
    "pathname": "/home",
    "search": "?test",
  },
  "type": "@@ROUTER/REPLACE",
}
`;
